luavm/src/lib/lua_values/mod.rs

```diff
@@ -0,0 +1,754 @@
+mod lua_table;
+mod tagging;
+
+use self::{lua_table::LuaTable, tagging::*};
+use errors::LuaError;
+use gc::{Finalize, Gc, Trace};
+
+/// Represents a value in Lua.
+#[derive(Debug)]
+pub struct LuaVal {
+    encoded_ptr: usize,
+    /// whether this should drop the pointer or not
+    owns: bool,
+}
+
+impl Finalize for LuaVal {}
+unsafe impl Trace for LuaVal {
+    custom_trace!(this, {
+        // only tables have a Gc inside, so there is no need to mark anything else
+        if let LuaValKind::TABLE = this.kind() {
+            mark(unsafe { &*table_ptr(this.encoded_ptr) });
+        }
+    });
+}
+
+impl LuaVal {
+    /// Create an empty LuaVal which is equivalent to Nil.
+    pub fn new() -> LuaVal {
+        LuaVal {
+            encoded_ptr: 0,
+            owns: true,
+        }
+    }
+
+    /// Copies the pointer into a new LuaVal.
+    fn shallow_copy(&self) -> LuaVal {
+        LuaVal {
+            encoded_ptr: self.encoded_ptr,
+            owns: false,
+        }
+    }
+
+    /// Returns the type of the value store in the pointer.
+    fn kind(&self) -> LuaValKind {
+        LuaValKind::from(self.encoded_ptr)
+    }
+
+    /// Returns true if this value can be considered a float, or false otherwise.
+    pub fn is_float(&self) -> bool {
+        match self.kind() {
+            LuaValKind::FLOAT | LuaValKind::STR => true,
+            _ => false,
+        }
+    }
+
+    /// Attempts to convert this value to a float.
+    pub fn to_float(&self) -> Result<f64, LuaError> {
+        unsafe {
+            match self.kind() {
+                LuaValKind::INT => Ok(*(i64_ptr(self.encoded_ptr)) as f64),
```

Comment:
This isn't safe: a large 64 bit int isn't representable as a 64 bit float. So I think you have to some sort of bounds checking or similar.

Reply:
It seems that `as f64` has this behaviour already, as far as I can tell.

Reply:
OK, then we should 1) document the expected behaviour 2) leave behind some `debug_assert_eq`s for the `as f64`. In particular we need to consider this:

> Casting from a float to an integer will round the float towards zero
>
>    NOTE: currently this will cause Undefined Behavior if the rounded value cannot be represented by the target integer type. This includes Inf and NaN. This is a bug and will be fixed.


https://doc.rust-lang.org/reference/expressions/operator-expr.html#type-cast-expressions

Reply:
What kind of asserts should I add? I am not quite sure what to test.

Reply:
Things like `debug_assert_eq!(<some very big float> as u64, ...)`.